gossip/src/contact_info.rs

```diff
@@ -0,0 +1,619 @@
+pub use crate::legacy_contact_info::LegacyContactInfo;
+use {
+    crate::crds_value::MAX_WALLCLOCK,
+    matches::debug_assert_matches,
+    serde::{Deserialize, Deserializer, Serialize},
+    solana_sdk::{
+        pubkey::Pubkey,
+        sanitize::{Sanitize, SanitizeError},
+        serde_varint, short_vec,
+    },
+    static_assertions::const_assert_eq,
+    std::{
+        collections::HashSet,
+        net::{IpAddr, Ipv4Addr, SocketAddr},
+        time::{SystemTime, UNIX_EPOCH},
+    },
+    thiserror::Error,
+};
+
+const SOCKET_TAG_GOSSIP: u8 = 0;
+const SOCKET_TAG_REPAIR: u8 = 1;
+const SOCKET_TAG_RPC: u8 = 2;
+const SOCKET_TAG_RPC_PUBSUB: u8 = 3;
+const SOCKET_TAG_SERVE_REPAIR: u8 = 4;
+const SOCKET_TAG_TPU: u8 = 5;
+const SOCKET_TAG_TPU_FORWARDS: u8 = 6;
+const SOCKET_TAG_TPU_FORWARDS_QUIC: u8 = 7;
+const SOCKET_TAG_TPU_QUIC: u8 = 8;
+const SOCKET_TAG_TPU_VOTE: u8 = 9;
+const SOCKET_TAG_TVU: u8 = 10;
+const SOCKET_TAG_TVU_FORWARDS: u8 = 11;
+const_assert_eq!(SOCKET_CACHE_SIZE, 12);
+const SOCKET_CACHE_SIZE: usize = SOCKET_TAG_TVU_FORWARDS as usize + 1usize;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Duplicate IP address: {0}")]
+    DuplicateIpAddr(IpAddr),
+    #[error("Duplicate socket: {0}")]
+    DuplicateSocket(/*key:*/ u8),
+    #[error("Invalid IP address index: {index}, num addrs: {num_addrs}")]
+    InvalidIpAddrIndex { index: u8, num_addrs: usize },
+    #[error("Invalid port: {0}")]
+    InvalidPort(/*port:*/ u16),
+    #[error("IP addresses saturated")]
+    IpAddrsSaturated,
+    #[error("Multicast IP address: {0}")]
+    MulticastIpAddr(IpAddr),
+    #[error("Port offsets overflow")]
+    PortOffsetsOverflow,
+    #[error("Socket not found: {0}")]
+    SocketNotFound(/*key:*/ u8),
+    #[error("Unspecified IP address: {0}")]
+    UnspecifiedIpAddr(IpAddr),
+    #[error("Unused IP address: {0}")]
+    UnusedIpAddr(IpAddr),
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, AbiExample, Serialize)]
+pub struct ContactInfo {
+    pubkey: Pubkey,
+    #[serde(with = "serde_varint")]
+    wallclock: u64,
+    // When the node instance was first created.
+    // Identifies duplicate running instances.
+    outset: u64,
+    shred_version: u16,
+    version: solana_version::Version,
+    // All IP addresses are unique and referenced at least once in sockets.
+    #[serde(with = "short_vec")]
+    addrs: Vec<IpAddr>,
+    // All sockets have a unique key and a valid IP address index.
+    #[serde(with = "short_vec")]
+    sockets: Vec<SocketEntry>,
+    #[serde(skip_serializing)]
+    cache: [SocketAddr; SOCKET_CACHE_SIZE],
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, AbiExample, Deserialize, Serialize)]
+struct SocketEntry {
+    key: u8,   // Protocol identifier, e.g. tvu, tpu, etc
+    index: u8, // IpAddr index in the accompanying addrs vector.
+    #[serde(with = "serde_varint")]
+    offset: u16, // Port offset with respect to the previous entry.
```

Comment:
I think the offset is not totally accurate, for the first entry it is the actual port -- so better to reflect it in the comment at least. I wonder what is the benefit of storing the offset than simply the port number itself as most complexity in this code is related to the offset arithmetic.

Reply:
Ports tend to be close to each other, so the offsets can be stored in one byte.
For gossip we are consuming a lot of bandwidth so anything which helps in the front can be useful.
Any added complexity is encapsulated inside this struct, and this is not exposed in the public api.